### PR TITLE
Fill in FLP overview in draft

### PR DESCRIPTION
**Pull request by junyechen1996**
The remaining two sections to fill:
- Wraparound checks and FLP circuit
- VDAF overview

### CODE REVIEW

File: draft-chen-cfrg-vdaf-pine.md [hunk 1]
```diff
@@ -139,10 +139,10 @@ functionality, but the concrete cost in terms of runtime and communication
 would be prohibitively high. The basic idea is simple: an FLP ("Fully Linear
 Proof", see {{Section 7.3 of !VDAF}}) could be specified that computes the
 L2-norm of the gradient and checks that the result is in the desired range.
-This computation is not easy to do efficiently: the challenge lies in ensuring that
```
Comment (junyechen1996) [inline draft-chen-cfrg-vdaf-pine.md:142]: Fix to wrap texts after 80 characters

File: draft-chen-cfrg-vdaf-pine.md [hunk 2]
```diff
@@ -222,13 +222,12 @@ does not wrap around the field modulus. However, this approach has relatively
 high communication overhead between the Client and Aggregators, roughly
 `num_frac_bits * dimension` field elements (see {{pine-user-param}}).
 
-In order to detect whether a wraparound has occurred, PINE uses a probabilistic test,
```
Comment (junyechen1996) [inline draft-chen-cfrg-vdaf-pine.md:225]: Fix to wrap texts after 80 characters

File: draft-chen-cfrg-vdaf-pine.md [hunk 3]
```diff
@@ -249,7 +249,9 @@ are chosen carefully in order to ensure this leakage is negligible.
 
 This section specifies a randomized encoding of gradients and FLP circuit
 ({{Section 7.3 of !VDAF}}) for checking that (1) the gradient's
-squared L2-norm falls in the desired range and (2) the squared L2-norm does not wrap around the field modulus.
+squared L2-norm falls in the desired range and (2) the squared L2-norm does
```
Comment (junyechen1996) [inline draft-chen-cfrg-vdaf-pine.md:252]: Fix to wrap texts after 80 characters

File: draft-chen-cfrg-vdaf-pine.md [hunk 4]
```diff
@@ -239,7 +238,8 @@ same distribution.
 
 However, {{ROCT23}} also show that each wraparound test has a non-zero false
 positive probability (the probability of misclassifying a valid gradient as
-invalid). This creates a problem for privacy, as the Aggregators learn
+invalid). We denote this probability as the "zero-knowledge error", or in short,
```
Comment (junyechen1996) [inline draft-chen-cfrg-vdaf-pine.md:241]: The goal is to briefly mention ZK error here.

File: draft-chen-cfrg-vdaf-pine.md [hunk 5]
```diff
@@ -185,7 +185,7 @@ The user-specified parameters to initialize PINE are defined in
 |:----------------|:------|:---------------|
 | `l2_norm_bound` | float | The L2-norm upper bound (inclusive). |
 | `dimension`     | int   | Dimension of each gradient. |
-| `num_frac_bits` | int   | The number of bits of precision to use when encoding each gradient entry into the field. At most XXX |
+| `num_frac_bits` | int   | The number of bits of precision to use when encoding each gradient entry into the field. At most 127. |
```
Comment (cjpatton) [inline draft-chen-cfrg-vdaf-pine.md:188]: The upper bound depends on the field and norm bound, correct?

Perhaps we should just say nothing here.
```suggestion
| `num_frac_bits` | int   | The number of bits of precision to use when encoding each gradient entry into the field. |
```

File: draft-chen-cfrg-vdaf-pine.md (hunk 4 above)
Comment (cjpatton) [inline draft-chen-cfrg-vdaf-pine.md:241]:
```suggestion
invalid). We refer to this probability as the "zero-knowledge error", or in short,
```

File: draft-chen-cfrg-vdaf-pine.md [hunk 6]
```diff
@@ -172,7 +172,7 @@ This document uses the same parameters and conventions specified for:
 
 * XOFs ("eXtendable Output Functions") from {{Section 6.2 of !VDAF}}.
 
-A floating point number is XXX. [CP: Specify, with a relvant reference.]
+A floating point number is a IEEE-754 compatible float64 value {{IEEE754-2019}}.
```
Comment (cjpatton) [inline draft-chen-cfrg-vdaf-pine.md:175]:
```suggestion
A floating point number, denoted `float`, is a IEEE-754 compatible float64 value {{IEEE754-2019}}.
```

File: draft-chen-cfrg-vdaf-pine.md [hunk 7]
```diff
@@ -259,40 +261,46 @@ derived from shares of the gradient.
...
 |:----------------------|:---------|:------------|
-| alpha                 | XXX      | XXX         |
-| num_wr_checks         | XXX      | XXX         |
-| num_wr_successes      | XXX      | XXX         |
-| encoded_sq_norm_bound | XXX      | XXX         |
-| wr_check_bound        | XXX      | XXX         |
-| num_bits_for_sq_norm  | XXX      | XXX         |
-| num_bits_for_wr_check | XXX      | XXX         |
-| chunk_len             | XXX      | XXX         |
+| alpha                 | float    | Parameter in wraparound check that determines the ZK error. The higher `alpha` is, the lower ZK error is. |
```
Comment (cjpatton) [inline draft-chen-cfrg-vdaf-pine.md:266]: Let's make sure to connect `float` to the spec (see suggestion above).

File: draft-chen-cfrg-vdaf-pine.md [hunk 8]
```diff
@@ -259,40 +261,46 @@ derived from shares of the gradient.
...
+| num_wr_checks         | int      | Number of wraparound checks to run. |
+| num_wr_successes      | int      | Minimum number of wraparound checks that a Client must pass. |
+| encoded_sq_norm_bound | Field    | The square of `l2_norm_bound` encoded into a field element. |
+| wr_check_bound        | Field    | The bound of the range check in wraparound checks. |
```
Comment (cjpatton) [inline draft-chen-cfrg-vdaf-pine.md:270]:
```suggestion
| wr_check_bound        | Field    | The bound of the range check for each wraparound check. |
```

File: draft-chen-cfrg-vdaf-pine.md [hunk 9]
```diff
@@ -259,40 +261,46 @@ derived from shares of the gradient.
...
+| num_bits_for_sq_norm  | int      | Number of bits to encode the squared L2-norm. |
+| num_bits_for_wr_check | int      | Number of bits to encode the range check in each wraparound check. |
+| chunk_length          | int      | Chunk length used in FLP proof. |
```
Comment (cjpatton) [inline draft-chen-cfrg-vdaf-pine.md:273]: The reader doesn't quite know yet what "chunk length" means
```suggestion
| chunk_length          | int      | Parameter of the FLP. |
```

File: draft-chen-cfrg-vdaf-pine.md [hunk 10]
```diff
@@ -259,40 +261,46 @@ derived from shares of the gradient.
...
-> XXX Give an overview of `encode_gradient()` and `encode_wr_checks()`. No need
-> to go into full detail yet. The goal is to give the reader an intuition for
-> what the circuit will do, including the bit checks, norm computation, and
-> wraparound checks. Be sure to emphasize that the wraparound test results are
-> not transmitted: instead the Aggregators derive shares of the tests results
-> by running the wraparound tests themselves.
-
-### Encoding of Floating Point Numbers into Field Elements
+The measurement encoding is done in two stages:
+* `PineValid.encode_gradient(self, measurement: list[float]) -> list[Field]`
```
Comment (cjpatton) [inline draft-chen-cfrg-vdaf-pine.md:279]: Don't refer to syntax that hasn't been defined yet. Instead, how about adding a forward reference to the section that describes it?
```suggestion
* 
```

File: draft-chen-cfrg-vdaf-pine.md [hunk 11]
```diff
@@ -259,40 +261,46 @@ derived from shares of the gradient.
...
+  field element {{float-to-field}}. Then the Client encodes the results for
+  L2-norm check {{l2-norm-check}}, by computing the bit representation of the
+  squared L2-norm, modulo `q`, of its encoded gradient. The result of this step
+  allows Aggregators to check the squared L2-norm of the Client's gradient,
+  modulo `q`, falls in the desired range of `[0, encoded_sq_norm_bound]`.
+* `PineValid.encode_wr_checks(self, encoded_gradient: list[Field],
+  wr_joint_rand_xof: Xof) -> list[Field]` allows the Client to encode the
+  results of wraparound checks {{wr-check}}, based on the `encoded_gradient`
+  from the previous step, and the random vector generated from an initialized
+  `Xof` in `wr_joint_rand_xof. The result of this step, along with the encoded
```
Comment (cjpatton) [inline draft-chen-cfrg-vdaf-pine.md:290]: Instead of speaking of an XOF here, we should only speak of the seed used to derive the random vectors via the XOF.

File: draft-chen-cfrg-vdaf-pine.md (hunk 10 above)
Comment (junyechen1996) [reply draft-chen-cfrg-vdaf-pine.md]: I added a section `Encoding Gradient {#encode-gradient}` and mentions this function there. I also put floating point encoding and L2-norm check as subsections of it.

File: draft-chen-cfrg-vdaf-pine.md (hunk 11 above)
Comment (junyechen1996) [reply draft-chen-cfrg-vdaf-pine.md]: isn't the XOF composed of the seed and domain separation tag? Mentioning XOF here also connects to our reference implementation?

File: draft-chen-cfrg-vdaf-pine.md (hunk 10 above)
Comment (cjpatton) [reply draft-chen-cfrg-vdaf-pine.md]: Works for me

File: draft-chen-cfrg-vdaf-pine.md [hunk 12]
```diff
@@ -259,45 +261,60 @@ derived from shares of the gradient.
...
+The measurement encoding is done in two stages:
+* {{encode-gradient}} involves encoding floating point numbers in the Client
+  gradient into field elements {{float-to-field}}, and encoding the results for
+  L2-norm check {{l2-norm-check}}, by computing the bit representation of the
+  squared L2-norm, modulo `q`, of the field element vector. The result of this
+  step allows Aggregators to check the squared L2-norm of the Client's gradient,
+  modulo `q`, falls in the desired range of `[0, encoded_sq_norm_bound]`.
+* {{encode-wr-check}} involves encoding the results of running wraparound checks
+  {{run-wr-check}}, based on the `encoded_gradient` from the previous step, and
+  the random vector generated from an initialized `Xof` in `wr_joint_rand_xof.
```
Comment (cjpatton) [inline draft-chen-cfrg-vdaf-pine.md:287]:
```suggestion
  the vectors derived from a short, random seed using an XOF.
````

**Review by cjpatton — APPROVED** [review]

**Comment by junyechen1996** [comment]
Squashed.
